src/main/java/com/minecolonies/api/crafting/RecipeStorage.java

```diff
@@ -252,12 +252,15 @@ private List<ItemStack> calculateSecondaryOutputs()
      */
     private void calculateTools()
     {
-        for(ItemStorage item : getCleanedInput())
+        final List<ItemStorage> cleaned = getCleanedInput();
+        for (int index = 0; index < cleaned.size(); index++)
         {
-            for(ItemStack result: getSecondaryOutputs())
+            final ItemStorage item = cleaned.get(index);
+            for (ItemStack result : getSecondaryOutputs())
             {
-                if(ItemStackUtils.compareItemStacksIgnoreStackSize(item.getItemStack(), result, false, true) && result.isDamageableItem())
+                if (ItemStackUtils.compareItemStacksIgnoreStackSize(item.getItemStack(), result, false, true) && result.isDamageableItem())
                 {
+                    cleaned.set(index, new ImmutableItemStorage(new ItemStorage(item.getItemStack(), item.getAmount(), true)));
```

Comment:
So, on calculate tools, you alter the cleanedInput?

Reply:
Yes, Mirality found that was the easiest fix to make this work

Reply:
Why don't we do a toolcheck in calculateCleanedInput()?

Reply:
You might as well remove the calculate tools function then

Reply:
We can technically do that indeed. But at the very least it fills up separate collections. But side injecting into a different one is weird.

Reply:
I tried to check it already and it's quite some dated/complex code, is it fine if we merge this first so the bug is solved, then look at improving the handling in this class?

Reply:
This side inject is too bug prone. Moving this specific code to the second loop in the calculateCleanedInput is straightforward

Reply:
I've tested it, it works fine as is, there are no bugs with it.
I made sure to turn it into a numeric for so you cannot have concurrent read/write problems that you get with generic fors.

Reply:
Yeah, I believe you tested it, but it's still much more bug prone this way because it the main input in an unexpected way for anyone else. Moving this to the calc simplifies this code by a massive amount

Reply:
The "side inject" to the cleaned input is not really any weirder than it altering the tools and secondary outputs -- it's just the thing it needs to do to do the job.  Both methods are only called from the constructor, but in different contexts (so simply moving the code to `calculateCleanedInput()` would not work).

What I would recommend (and might make you happier about it) is to s/`getCleanedInput()`/`cleanedInput`/ in this method -- it's a lot less weird to alter a field directly than to alter the result of a getter method, even if in this case they happen to be equivalent.